src/Microsoft.Health.Dicom.Core/Features/Update/UpdateInstanceService.cs

```diff
@@ -0,0 +1,193 @@
+    {
+        long originFileIdentifier = instanceFileIdentifier.Version;
+        long newFileIdentifier = instanceFileIdentifier.NewVersion.Value;
+
+        _logger.LogInformation("Begin downloading original file metdata {OrignalFileIdentifier} - {NewFileIdentifier}", originFileIdentifier, newFileIdentifier);
+
+        DicomDataset metadata = await _metadataStore.GetInstanceMetadataAsync(originFileIdentifier, cancellationToken);
+        metadata.AddOrUpdate(datasetToUpdate);
+
+        await _metadataStore.StoreInstanceMetadataAsync(metadata, newFileIdentifier, cancellationToken);
+
+        _logger.LogInformation("Updating metadata file {OrignalFileIdentifier} - {NewFileIdentifier} completed successfully", originFileIdentifier, newFileIdentifier);
+    }
+
+    private async Task UpdateDatasetInFileAsync(long newFileIdentifier, DicomDataset datasetToUpdate, KeyValuePair<string, long> block = default, CancellationToken cancellationToken = default)
+    {
+        _logger.LogInformation("Begin updating new file {NewFileIdentifier}", newFileIdentifier);
+
+        // If the block is not provided, then we need to get the first block to update the dataset
+        if (block.Key is null)
+        {
+            block = await _fileStore.GetFirstBlockPropertyAsync(newFileIdentifier, cancellationToken);
+        }
+
+        BinaryData data = await _fileStore.GetFileInRangeAsync(newFileIdentifier, new FrameRange(0, block.Value), cancellationToken);
+
+        using MemoryStream stream = _recyclableMemoryStreamManager.GetStream(data);
+        DicomFile dicomFile = await DicomFile.OpenAsync(stream);
+        dicomFile.Dataset.AddOrUpdate(datasetToUpdate);
+
+        using MemoryStream resultStream = _recyclableMemoryStreamManager.GetStream();
+        await dicomFile.SaveAsync(resultStream);
+
+        await _fileStore.UpdateFileBlockAsync(newFileIdentifier, block.Key, resultStream, cancellationToken);
+
+        _logger.LogInformation("Updating new file {NewFileIdentifier} completed successfully", newFileIdentifier);
+    }
+
+    private static IDictionary<string, long> GetBlockLengths(long streamLength, long initialBlockLength)
+    {
+        var blockLengths = new Dictionary<string, long>();
+        long fileSizeWithoutFirstBlock = streamLength - initialBlockLength;
+        int numStagesWithoutFirstBlock = (int)Math.Ceiling((double)fileSizeWithoutFirstBlock / StageBlockSizeInBytes) + 1;
+
+        long bytesRead = 0;
+        for (int i = 0; i < numStagesWithoutFirstBlock; i++)
+        {
+            long blockSize = i == 0 ? initialBlockLength : Math.Min(StageBlockSizeInBytes, streamLength - bytesRead);
+            string blockId = Convert.ToBase64String(Guid.NewGuid().ToByteArray());
+            blockLengths.Add(blockId, blockSize);
+            bytesRead += blockSize;
+        }
+
+        return blockLengths;
+    }
+
+    private static void RemoveItemsFromDataset(DicomDataset dataset, DicomItem largeDicomItem)
```

Comment:
Yes, renamed to RemoveItemsAfter. Whether pixeldata or non pixel data, if it is more than a certain length we would like to remove so that we can push the first set of items in the dataset to first block